src/dma/stream.rs

```diff
@@ -181,340 +181,344 @@ impl<I: Instance> StreamsTuple<I> {
-                                    .$gif().set_bit() //Clear global interrupt flag
-                    );
-                    let _ = dma.isr.read();
-                    let _ = dma.isr.read(); // Delay 2 peripheral clocks
-                }
-
-                #[inline(always)]
-                fn clear_transfer_complete_flag(&mut self) {
-                    //NOTE(unsafe) Atomic write with no side-effects and we only access the bits
-                    // that belongs to the StreamX
-                    let dma = unsafe { &*I::ptr() };
-                    dma.ifcr.write(|w| w.$tcif().set_bit());
-                }
+    ($(($name:ident, $number:expr, $x:literal)),+$(,)*) => {
+        paste::paste!{
+            $(
+                impl<I: Instance> Stream for $name<I> {
+
+                    const NUMBER: usize = $number;
+                    type Config = DmaConfig;
+                    type Interrupts = DmaInterrupts;
+
+                    fn apply_config(&mut self, config: DmaConfig) {
+                        self.set_priority(config.priority);
+                        self.set_memory_increment(config.memory_increment);
+                        self.set_peripheral_increment(config.peripheral_increment);
+                        self.set_circular_buffer(config.circular_buffer);
+                        self.set_transfer_complete_interrupt_enable(
+                            config.transfer_complete_interrupt
+                        );
+                        self.set_half_transfer_interrupt_enable(config.half_transfer_interrupt);
+                        self.set_transfer_error_interrupt_enable(
+                            config.transfer_error_interrupt
+                        );
+                    }
 
-                #[inline(always)]
-                fn clear_transfer_complete_interrupt(&mut self) {
-                    self.clear_transfer_complete_flag();
-                    //NOTE(unsafe) Atomic read with no side-effects.
-                    let dma = unsafe { &*I::ptr() };
-                    let _ = dma.isr.read();
-                    let _ = dma.isr.read(); // Delay 2 peripheral clocks
-                }
+                    #[inline(always)]
+                    fn clear_interrupts(&mut self) {
+                        //NOTE(unsafe) Atomic write with no side-effects and we only access the bits
+                        // that belongs to the StreamX
+                        let dma = unsafe { &*I::ptr() };
+                        dma.ifcr().write(|w| w
+                                        .[<ctcif $x>]().set_bit() //Clear transfer complete interrupt flag
+                                        .[<chtif $x>]().set_bit() //Clear half transfer interrupt flag
+                                        .[<cteif $x>]().set_bit() //Clear transfer error interrupt flag
+                                        .[<cgif $x>]().set_bit() //Clear global interrupt flag
+                        );
```

Comment:
DMA can be fully demacrofied:
```suggestion
                        let num = Self::NUMBER as u8;
                        dma.ifcr().write(|w|{
                            w.ctcif(num).set_bit(); //Clear transfer complete interrupt flag
                            w.chtif(num).set_bit(); //Clear half transfer interrupt flag
                            w.cteif(num).set_bit(); //Clear transfer error interrupt flag
                            w.cgif(num).set_bit() //Clear global interrupt flag
                        });
```